src/ocean/util/app/Application.d

```diff
@@ -334,13 +334,16 @@ class Application : IApplication
 
     /***************************************************************************
 
-        Default application extension order.
+        Default application extension order. It is set to very high value as
+        it is convenient for applications to assume their methods run last
+        but still leaves the space for rare few extension to set even higher
+        order.
 
     ***************************************************************************/
 
     public override int order ( )
     {
-        return 0;
+        return 1_000_000;
```

Comment:
I don't like it. I like that the application has order 0, is the "center" of the framework. Why changing this at all?

Reply:
I see there are a lot of extensions using order 0, I think the correct fix would be to make those extensions use some negative value that makes sense (here is when it gets tricky, to get an order that make the extensions play well between themselves).

Reply:
@leandro-lucarella-sociomantic we could adopt the BASIC line numbering principle of increments of 10 or more, to leave room for other priorities to be inserted ... :trollface: 

Note that 0 is explicitly described in the `IExtension` doc as the default "if order doesn't matter" setting:
https://github.com/sociomantic-tsunami/ocean/blob/v4.1.1/src/ocean/util/app/model/IExtension.d#L35-L36

So if this is true, it is _not_ the case that the correct `order` for the app extension is 0, because for the app extension the order _does_ matter.

A related question is if it can possibly be true (or reasonable) that an extension could/should have a higher `order` than the application extension.

Reply:
Note, I'm not anti the idea of 0 being reinterpreted as application-extension priority, so individual extensions should specify priority in line with whether they need to be ready before or after the application extension.  But that means we have to change the meaning of the `order` return value.

Reply:
Reading the docs of `Application.d` it looks like we have a clash here between what is specified in `IExtension.order` doc and what is specified in `Application` doc.

Reply:
> I don't like it. I like that the application has order 0, is the "center" of the framework. Why changing this at all?

Because `0` is the default value for extension order and we need application to have lower order by default. And relying on virtue of developers to always modify order value of new added extensions is not going to work.

We can keep app at 0 if base `order` method return value for all extensions is defined to `-1_000` or something like that.

Reply:
There is no default, you have to always explicitly set one. To be honest, I can understand that maybe the docs are difficult to navigate, but it's all there. I think this is a docs-only issue, so maybe we can do that, recommend some value for extensions (like `-1_000`), saying that extensions want to almost always run before the application.

Reply:
That sounds exactly the same as setting app to `1_000` but (considerably) more error-prone. Why is this so important? Application framework is convoluted enough that at this point you can safely assume no on reads docs when writing extensions, thus suitable defaults are extremely important.

Reply:
It's not important but minimizes the chances that we fuck up something else. For example your PR didn't update the Application docs. There might be other areas where the App having order of 0 is assumed. I don't see a strong point to change it, and without a strong point, I prefer to minimize the chances of fuck ups.

Reply:
Specially for a patch-release for me this makes NO sense. If it were a major (or maybe a minor) then maybe. I don't know if other apps might be relying on this order.

Reply:
Right now it is semi-undefined (in a sense that most extensions use order 0 and resulting order ends up being "reverse order of register calls") - hard to rely on this.

Will you be OK with having this in a minor?

Reply:
> Right now it is semi-undefined (in a sense that most extensions use order 0 and resulting order ends up being "reverse order of register calls") - hard to rely on this.

Well, for those extensions order was important, so I would say is a bug in those extensions and those extensions should be fixed (change the order to some negative value). For me that's the fix that should go in a patch release.

> Will you be OK with having this in a minor?

Change the Application order to something bigger than 0? I'll be OK-ish, but I strongly would recommend against it, because is a silent change on behaviour for any project deriving from `Application` that could have unpredictable and hard to track consequences. I think that sort of changes belong to a major.

Reply:
@gavin-norman-sociomantic it would be useful to have your POV on this.

Reply:
Given that:
1. I think we should seriously look into a major refactoring of the internals of the app framework soon,
2. it's pretty unlikely that more extensions will be added any time soon,

I think it makes the most sense to keep the meaning of 0 as the natural middle point (app initialisation), and move all extensions that should happen before then to a negative order.